docs/notebooks/mixed_search_spaces.pct.py

```diff
@@ -0,0 +1,209 @@
+#
+# The Branin function is a common optimization benchmark that has three global minima. It is normally
+# defined over a 2D continuous search space.
+#
+# We first show the Branin function over its original continuous search space.
+
+# %%
+from trieste.experimental.plotting import plot_function_plotly
+from trieste.objectives import ScaledBranin
+
+scaled_branin = ScaledBranin.objective
+
+fig = plot_function_plotly(
+    scaled_branin,
+    ScaledBranin.search_space.lower,
+    ScaledBranin.search_space.upper,
+)
+fig.show()
+
+# %% [markdown]
+# In order to convert the Branin function from a continuous to a mixed search space problem, we
+# modify it by discretizing its first input dimension.
+#
+# The discrete dimension is defined by a set of 10 points that are equally spaced, ensuring that
+# the three minimizers are included in this set. The continuous dimension is defined by the
+# interval [0, 1].
+#
+# We observe that the first and third minimizers are equidistant from the middle minimizer, so we
+# choose the discretization points to be equally spaced around the middle minimizer.
+
+# %%
+from trieste.space import Box, DiscreteSearchSpace, TaggedProductSearchSpace
+
+minimizers0 = ScaledBranin.minimizers[:, 0]
+step = (minimizers0[1] - minimizers0[0]) / 4
+points = np.concatenate(
+    [
+        # Equally spaced points to the left of the middle minimizer. Skip the last point as it is
+        # the same as the first point in the next array.
+        np.flip(np.arange(minimizers0[1], 0.0, -step))[:-1],
+        # Equally spaced points to the right of the middle minimizer.
+        np.arange(minimizers0[1], 1.0, step),
+    ]
+)
+discrete_space = DiscreteSearchSpace(points[:, None])
+continuous_space = Box([0.0], [1.0])
+mixed_search_space = TaggedProductSearchSpace(
+    [discrete_space, continuous_space]
+)
+
+
+# %% [markdown]
+# Next we demonstrate the Branin function over the mixed search space, by plotting the original
+# function contours and highlighting the discretization points.
+# The discrete dimension is along the x-axis and the continuous dimension is on the y-axis, with
+# the vertical dashed lines indicating the discretization points.
+
+# %%
+from trieste.experimental.plotting import plot_function_2d
+
+# Plot over the predefined search space.
+fig, ax = plot_function_2d(
+    scaled_branin,
+    ScaledBranin.search_space.lower,
+    ScaledBranin.search_space.upper,
+    contour=True,
+)
+
+ax[0, 0].set_xlabel(r"$x_1$")
+ax[0, 0].set_ylabel(r"$x_2$")
+
+# Draw vertical lines at the discrete points.
+for point in points:
+    ax[0, 0].vlines(
+        point,
+        mixed_search_space.lower[1],
+        mixed_search_space.upper[1],
+        colors="b",
+        linestyles="dashed",
+        alpha=0.6,
+    )
+
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# We begin our optimization by first collecting five function evaluations from random locations in
+# the mixed search space.
```

Comment:
Add something like "observe how the sample method deals with the mixed space automatically blabla"?

(maybe we should be explicit that for discrete spaces it will be sampling with replacement)